workflow/controller/controller.go

```diff
@@ -113,39 +105,107 @@ func (wfc *WorkflowController) Run(ctx context.Context) error {
 		return err
 	}
 
-	log.Info("Watch Workflow objects")
+	wfc.wfInformer = wfc.newWorkflowInformer()
+	wfc.podInformer = wfc.newPodInformer()
+	go wfc.wfInformer.Run(ctx.Done())
+	go wfc.podInformer.Run(ctx.Done())
 
-	// Watch Workflow objects
-	_, err = wfc.watchWorkflows(ctx)
-	if err != nil {
-		log.Errorf("Failed to register watch for Workflow resource: %v", err)
-		return err
+	// Wait for all involved caches to be synced, before processing items from the queue is started
+	for _, informer := range []cache.SharedIndexInformer{wfc.wfInformer, wfc.podInformer} {
+		if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
+			return errors.InternalError("Timed out waiting for caches to sync")
+		}
+	}
+
+	wfc.StartStatsTicker(5 * time.Minute)
+	for i := 0; i < 4; i++ {
```

Comment:
Does this mean there are 4 worker threads/goroutines for processing workflows? And below on line 124, there are 8 threads/goroutines for processing pods?

Reply:
Yes, this is something that needs to be tuned, but just arbitrarily picked those for now.

Reply:
Let's file bugs for known improvements so that it isn't forgotten.

Reply:
Issue #568 filed.